### PR TITLE
[FLINK-6904] [cep] Support for quantifier range to CEP's pattern API

### PR DESCRIPTION
**Pull request by dianfu**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [x] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFAITCase.java [hunk 1]
```diff
@@ -1391,6 +1391,248 @@ public boolean filter(Event value) throws Exception {
 	}
```
Comment (dawidwys) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFAITCase.java]: Maybe let's move all those new tests into seperate class like `SameElementITCase` or `IterativeConditionsITCase`? The NFAITCase is already too long in my opinion.

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFAITCase.java [hunk 2]
```diff
@@ -1905,6 +2147,275 @@ public boolean filter(Event value) throws Exception {
 		));
 	}
 
+	@Test
+	public void testTimesRangeNonStrictOptional1() {
```
Comment (dawidwys) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFAITCase.java]: I think it is covered by the next two tests.

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFAITCase.java [hunk 3]
```diff
@@ -1905,6 +2147,275 @@ public boolean filter(Event value) throws Exception {
...
+
+		compareMaps(resultingPatterns, Lists.<List<Event>>newArrayList(
+			Lists.newArrayList(ConsecutiveData.startEvent, ConsecutiveData.middleEvent1, ConsecutiveData.middleEvent2, ConsecutiveData.middleEvent3, ConsecutiveData.end),
+			Lists.newArrayList(ConsecutiveData.startEvent, ConsecutiveData.middleEvent1, ConsecutiveData.middleEvent2, ConsecutiveData.end),
+			Lists.newArrayList(ConsecutiveData.startEvent, ConsecutiveData.middleEvent1, ConsecutiveData.middleEvent3, ConsecutiveData.end)
+		));
+	}
+
+	@Test
+	public void testTimesRangeNotStrictWithFollowedByEager() {
```
Comment (dawidwys) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFAITCase.java]: Maybe change the name to `testTimesRangeNotStrictWithFollowedBy`. I think eager may be misleading here. What do you think?

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFAITCase.java [hunk 4]
```diff
@@ -1905,6 +2147,275 @@ public boolean filter(Event value) throws Exception {
...
+		List<List<Event>> resultingPatterns = feedNFA(inputEvents, nfa);
+
+		compareMaps(resultingPatterns, Lists.<List<Event>>newArrayList(
+			Lists.newArrayList(ConsecutiveData.startEvent, ConsecutiveData.middleEvent1, ConsecutiveData.middleEvent2, ConsecutiveData.middleEvent3, ConsecutiveData.end),
+			Lists.newArrayList(ConsecutiveData.startEvent, ConsecutiveData.middleEvent1, ConsecutiveData.middleEvent2, ConsecutiveData.end)
+		));
+	}
+
+	@Test
+	public void testTimesRangeNotStrictWithFollowedByNotEager() {
```
Comment (dawidwys) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFAITCase.java]: Maybe change the name to `testTimesRangeNotStrictWithFollowedByAny`. I think eager may be misleading here. What do you think?

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFAITCase.java (hunk 1 above)
Comment (dianfu) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFAITCase.java]: Agree. Moved the newly added tests to file TestsRangeITCase.

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFAITCase.java (hunk 2 above)
Comment (dianfu) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFAITCase.java]: The next two tests are slightly different from this one:
testTimesRangeNotStrictOption1: followedBy().times().optional().followedBy
testTimesRangeNotStrictOption2: followedByAny().times().allowCombinations().optional().followedBy()
testTimesRangeNotStrictOption3: followedByAny().times().optional().followedBy()

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFAITCase.java (hunk 3 above)
Comment (dianfu) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFAITCase.java]: Agree, addressed in the new patch.

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFAITCase.java (hunk 4 above)
Comment (dianfu) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFAITCase.java]: Agree, addressed in the new patch.

**Comment by dianfu** [comment]
Hi @dawidwys thanks a lot for the review, what's your thought for the updated patch?

**Comment by dawidwys** [comment]
Hi @dianfu, sorry for late reply. I was on short vacations. The updated PR looks good! I agree though with @kl0u that changes from #4132 should be covered by this PR.

**Comment by dianfu** [comment]
@dawidwys make sense, have updated the patch.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/pattern/Pattern.java [hunk 5]
```diff
@@ -318,7 +319,30 @@ public Quantifier getQuantifier() {
...
+	 * @param from number of times matching event must appear at least
+	 * @param to number of times matching event must appear at most
+	 * @return The same pattern with the number of times range applied
+	 *
+	 * @throws MalformedPatternException if the quantifier is not applicable to this pattern.
+	 */
+	public Pattern<T, F> times(int from, int to) {
+		checkIfNoNotPattern();
+		checkIfQuantifierApplied();
+		Preconditions.checkArgument(from >= 0, "The from should be a non-negative number greater than or equal to 0.");
```
Comment (dawidwys) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/pattern/Pattern.java]: It is already checked in `Times` class. Maybe remove those checks from here?

**Comment by dawidwys** [comment]
Thanks a lot @dianfu for the work here. I had the last comment and will merge it as soon as travis gives green.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/pattern/Pattern.java (hunk 5 above)
Comment (dianfu) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/pattern/Pattern.java]: Make sense, removed.

**Pull request closed without merging**